mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/memory/DeleteByQueryExecutor.java

```diff
@@ -0,0 +1,92 @@
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class DeleteByQueryExecutor {
+    static class Notifiers {
+        private final Runnable userHandledNotifier;
+        private final Runnable searchErrorNotifier;
+        private final Runnable deletionErrorNotifier;
+        private final Runnable permanentlyDeletedMessageNotifyer;
+
+        Notifiers(Runnable userHandledNotifier, Runnable searchErrorNotifier, Runnable deletionErrorNotifier, Runnable permanentlyDeletedMessageNotifyer) {
+            this.userHandledNotifier = userHandledNotifier;
+            this.searchErrorNotifier = searchErrorNotifier;
+            this.deletionErrorNotifier = deletionErrorNotifier;
+            this.permanentlyDeletedMessageNotifyer = permanentlyDeletedMessageNotifyer;
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeleteByQueryExecutor.class);
+
+    private final DeletedMessageVault deletedMessageVault;
+
+    public DeleteByQueryExecutor(DeletedMessageVault deletedMessageVault) {
+        this.deletedMessageVault = deletedMessageVault;
+    }
+
+    public Task.Result deleteByQuery(Query query, Notifiers notifiers) {
+        return Flux.from(deletedMessageVault.usersWithVault())
+            .flatMap(user -> deleteByQueryForUser(query, user, notifiers))
+            .reduce(Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Unexpected error encountered while deleting by query", e);
+                return Mono.just(Task.Result.PARTIAL);
+            })
+            .blockOptional()
+            .orElse(Task.Result.COMPLETED);
+    }
+
+    private Mono<Task.Result> deleteByQueryForUser(Query allBeforeRetentionPeriod, User user, Notifiers notifiers) {
+        return Flux.from(deletedMessageVault.search(user, allBeforeRetentionPeriod))
+            .flatMap(message -> deleteMessage(user, message.getMessageId(), notifiers))
+            .onErrorResume(e -> {
+                LOGGER.error("Error encountered while searching old mails in {} vault", user.asString(), e);
+                notifiers.searchErrorNotifier.run();
+                return Mono.just(Task.Result.PARTIAL);
+            })
+            .reduce(Task::combine)
+            .map(FunctionalUtils.identityWithSideEffect(() -> LOGGER.info("Retention applied for {} vault", user.asString())))
+            .map(FunctionalUtils.identityWithSideEffect(notifiers.userHandledNotifier));
+    }
+
+    private Mono<Task.Result> deleteMessage(User user, MessageId messageId, Notifiers notifiers) {
+        return Mono.from(deletedMessageVault.delete(user, messageId))
+            .then(Mono.fromRunnable(notifiers.permanentlyDeletedMessageNotifyer)
+                .then(Mono.just(Task.Result.COMPLETED)))
```

Comment:
why do you need a level of nesting here?

Reply:
If you find an other equivalent of `Stream::peek` that is still synchronous, I take it!